### ISSUE TITLE
APSIM Classic - process for continuous release

### ISSUE THREAD

**Issue by sarahcleary** [APSIMInitiative/ReferencePanel#15]
For the forseeable future, the Apsim Classic release will remain at 7.10, and links to upgrade to the latest bugfix are available to developers only via the bootleg system on bob. This means registered users don't have a) knowledge of available bug fixes, and b) access to them. 

@peter-devoil to work with @hol353 to propose a solution which is likely to involve version and build number. Once agreed, @sarahcleary to provide a news article and web page for process and referencing.

**Comment by sarahcleary** [comment]
Any update @peter-devoil / @hol353 ?

**Comment by sarahcleary** [comment]
Any update @peter-devoil / @hol353 ?

**Comment by peter-devoil** [comment]
Bob has been broken and unable to create releases, so nothing has happened.

**Comment by sarahcleary** [comment]
Any update @peter-devoil / @hol353 ?

**Comment by peter-devoil** [comment]
Still waiting on bob - all the PRs seem to be failing. eg https://github.com/APSIMInitiative/APSIMClassic/pull/1746

**Comment by sarahcleary** [comment]
@peter-devoil - any update?

**Comment by sarahcleary** [comment]
@peter-devoil - any update?

**Comment by sarahcleary** [comment]
relates to https://github.com/APSIMInitiative/APSIMClassic/pull/1759  
@peter-devoil - can this issue be closed?

**Comment by peter-devoil** [comment]
No, still waiting on bob to a) stamp sequence numbers on the release, and b) build things.. See https://github.com/APSIMInitiative/APSIMClassic/issues/1757

**Comment by sarahcleary** [comment]
@peter-devoil - did you want this on the agenda for next Friday's meeting?  any update?

**Comment by peter-devoil** [comment]
@sarahcleary - Status is unchanged, but minor progress on the builder. Try again next time.

**Comment by sarahcleary** [comment]
@peter-devoil - did you want this on the agenda for Thursday's meeting? any update?

**Comment by peter-devoil** [comment]
Yes, there is a minor update in that - @hol430 has got jenkins building releases again, and we now have to work out how these installation files will appear on the downloads (upgrades?) section of the website.

**Comment by hol430** [comment]
They're currently accessible on [Bob's webpage](https://apsimdev.apsim.info/APSIM.Builds.Portal/Bob.aspx). Do we want them somewhere else?

**Comment by peter-devoil** [comment]
Yes, we want a simplified list for ordinary users with a) a link to the issue, b) a link to the installer, and c) the standard disclaimer. The above has works in progress and failures - users don't want to see that - the user just wants to know of the issue that was resolved, and a link to install the fix.

I don't know whether the downloads section of apsim.info should point to an autogenerated apsimdev page; or whether @sarahcleary should periodically update a page in the WP site. Given that we only resolved ~6 bugs in apsim classic last year, it might be simpler to do it by hand.

**Comment by peter-devoil** [comment]
I forgot to add d) sequence numbers. They need to start somewhere after the last svn revision number so that users don't think they're going backwards. 

I suspect @sarahcleary should be able to use something like [this wp plugin](https://wordpress.org/plugins/remote-content-shortcode/) on the main site somewhere that points to apsimdev - but I don't know whether the laborious registration procedure has to be followed?

@jbrider ~~@hol353~~ @hol430  - can we meet sometime in january to sort out what has to happen here?

**Comment by jbrider** [comment]
@peter-devoil I will be available after the 13th.
I think we're going to need incrementing version numbers for both classic and nextgen.
We've already had concerns voiced by developers about the versioning being tied to the issue number - it's only going to get worse when we start getting more regular users after the symposium.

**Comment by sarahcleary** [comment]
@jbrider; @peter-devoil; @hol430  - please let me know what I need to do with respect to the website.

**Comment by peter-devoil** [comment]
@sarahcleary - I think adding bits to apsim.info will be the last step in this process. But you can clear up a potential hazard for us: do users wanting to use the updated, latest apsim have to agree to the T&C they've already agreed to? @hol430 - if we ask the user to enter their email address, can we query the registrations DB to see if a user has agreed in the past?

**Comment by hol430** [comment]
Yeah that would be possible. How would this work if they've already agreed? On the website, the "download" button (if you can call it that) is already called something like "I accept the license" so we may not need changes there?

**Comment by peter-devoil** [comment]
I'd imagine that you'd start with an empty page and an "Enter your email address (the one you used when registering)" input box. Once they've entered a valid email, and it exists in the DB (make sure the email string is scrubbed for SQL injection), a list of downloads would appear. These would be links to the successful merges from [Bob](https://apsimdev.apsim.info/APSIM.Builds.Portal/Bob.aspx) - each line would have a) a link to the (github) issue that was resolved, and b) a download link to the installer. 

Don't worry about the other links - they're only interesting to nerds.

**Comment by hol430** [comment]
So we want to provide the users with a direct link immediately - rather than emailing them the link?

**Comment by peter-devoil** [comment]
What do you think, @sarahcleary ? Is it better for download links to be on the website, or in an email?

**Comment by sarahcleary** [comment]
@peter-devoil - can we have a quick discussion at the RP meeting tomorrow?  Download links should not be on the website - unless there is a way that a 'valid-registered' email address is entered - ie a way to check a user has agreed to the T&Cs (preferably within the time frame of the agreement they 'agreed to'.

**Comment by peter-devoil** [comment]
(I think) It was decided at the RP meeting today that we will provide download links on the website, available to users who have previously agreed to the T&C (ie their address is in the DB). @hol430, could you liase with sarah's WP expert to make a cgi script that does this?

**Comment by hol430** [comment]
Do we want some mechanism to validate the email addresses?

**Comment by jbrider** [comment]
The hope was that the email address of registered users was available in a
database.
To get into that database, it should have been verified already? not sure
if this is how it works.

Then when they went to a download page, they would have to enter an email
address (hopefully it is their own), and if it is in the database, show a
list of currently available version for download.
The list would be like the list on bob, but only show the executables
available for each successful build.

**Comment by hol430** [comment]
Currently, there is no validation of email addresses. That being said, the only way to obtain a download link is to receive the registration email (or just have a peek in the source code), so there's not really much point providing a fake email address.

**Comment by jbrider** [comment]
Are they available in a database then - to verify when the user enters one?
The idea is to remind them about the Terms & Conditions by making them
enter an email address before seeing the list.

**Comment by sarahcleary** [comment]
@hol430 - is this possible?  ie to 'validate' the email address prior to a user accessing the links via the website.  @jbrider thought it should be.  We could approach Andrew to see how best to do this on wordpress.

**Comment by hol430** [comment]
Yes, it ought to be possible. I'm just not sure it would end up looking much different to the current system for new users. I guess for existing users, it will be much more convenient.

**Comment by peter-devoil** [comment]
The database is referenced [here](https://github.com/APSIMInitiative/APSIM.Registration/blob/master/APSIM.Registration.Service/Registration.cs).

Yes, new users still agree to T&Cs, and after that it's simple to upgrade.

**Comment by sarahcleary** [comment]
@hol430 - @peter-devoil and I are meeting in person tomorrow, would you like us to call you to discuss?

**Comment by hol430** [comment]
@peter-devoil - do we provide macOS binaries for apsim classic?

**Comment by peter-devoil** [comment]
No. Mac users need a windows VM.

**Comment by hol430** [comment]
Wait, so you can't even do a non-GUI run on Mac?

Also, what would be the process for a brand-new user? Do they get emailed a download link? Do they get sent a verification email? Or do they just get redirected immediately to the downloads page without any verification?

**Comment by peter-devoil** [comment]
Mac users use [Parallels](https://www.parallels.com/au/products/desktop/) (or similar) to run the GUI.

I dont think we're changing anything for the brand new user - they register and get emailed a link to 7.10 the same as now - the last "official" release.

Later, they visit the website to check for updates, which lists the subsequent builds since the official release. The only verification at that point is that they've supplied an email address that they used to register apsim once upon a time..

**Comment by hol430** [comment]
Makes sense. Thanks

**Comment by hol430** [comment]
The ~HowOften~HowMuch installer doesn't run for me on Windows 10. Is there some dependency which I need to install?

![image](https://user-images.githubusercontent.com/36427516/74614523-3b330400-5164-11ea-80eb-75d8dc3bd994.png)

Edit: HowMuch, not HowOften.

**Comment by peter-devoil** [comment]
Probably the dependancy is windows XP.. All that stuff (HowXXX) is legacy now, we should ask  info@howleaky.net & [co](https://howleaky.net/index.php/contact/1-david-mcclymont) whether it's still appropriate for us to keep the links alive.

**Comment by hol430** [comment]
Well the other HowXXX programs apart from PERFECT seem to work, although the extent of my testing is "does the GUI start". I will send an email and find out whether they still want us to keep those links alive.

Anyway, I have a working prototype of this downloads page, as well as a process to take existing users directly to the downloads page, rather than forcing them to re-register. However I think it may be best if we wait until after apsim week to deploy this. Maybe that way I can show you guys in person as well while we're in Brisbane - the UI is currently a bit bare-bones and could probably use some tlc.

@sarahcleary fyi

**Comment by Keith-Pembleton** [comment]
I recall in an RP meeting last year we decided to to stop supporting the HowXXX stuff.  
BTW the new website for HowLeaky (and presumably contacts) is http://www.howleaky.com/home

**Comment by sarahcleary** [comment]
That is correct.  I did not transfer these onto the new website given the discussions of the RP.

**Comment by sarahcleary** [comment]
this is still outstanding.   We had intended on completing in person during APSIM week -  @peter-devoil, @jbrider @hol430 - suggestions on best way to complete?  teleconference?

**Comment by hol430** [comment]
I have a working prototype of this but it doesn't include the extra options for commercial users. I will add this in and then see where we're at.

**Comment by hol430** [comment]
How do you want this commercial license registration to work? I can create a separate webpage for commercial users but if many of the inputs are the same as for regular users (name, company, country, ...) then it might make more sense to use the same page. I can add a checkbox for "Download under commercial license" and by checking the checkbox, some extra inputs appear.

**Comment by sarahcleary** [comment]
@hol430 - I can't see why the same page, with a 'checkbox' can't work.  As long as we can pull out the list of commercial licensees easily.  @hol353; @peter-devoil  - can you see any issues doing it this way?

**Comment by sarahcleary** [comment]
@hol430  - just to confirm - the prototype you talk about above, combines both issues - the one for registration and the one to allow APSIM Classic cts release.  @peter-devoil - anything you wish to add/clarify?

**Comment by hol430** [comment]
What does cts stand for?

**Comment by sarahcleary** [comment]
cts - continuous

**Comment by peter-devoil** [comment]
Nothing from me - happy to wait till we get a prototype.

**Comment by hol430** [comment]
Ok - yeah, I've implemented the changes to the registration page that you wanted - although I haven't made these changes live yet. I also haven't done anything about the modified registration process for commercial users as I don't know what extra information you want. That can be considered a separate issue/change though - we can proceed with just the revamped registration process for now.

![image](https://user-images.githubusercontent.com/36427516/78327184-9f4c4300-75bf-11ea-970b-e70c9f31806c.png)

![image](https://user-images.githubusercontent.com/36427516/78327290-d884b300-75bf-11ea-8885-ecbe9b3fe3d2.png)

I haven't implemented an incrementing version number yet. @peter-devoil how did this work in the past? It seems to me that whenever a pull request gets merged, we would have to retest all of the other pull requests? Otherwise the version info embedded in the build artifacts would be stale.

**Comment by peter-devoil** [comment]
It was awful. Basically, every time a build was made, it ran with a candidate revision number - current + 1 . If it was accepted (passed automated tests) the build was good to go. If it failed, the installers are discarded and the build is only used for dissection. 

It's getting it wrong here - the 7.7r0 bits. There was a field in the database for this iirc. 

At the worst, it might mean having to do the build twice - once for testing, another for branding (with rXXXX) after the merge. I suspect we could get away without a full build, as its only the Release scripts (that create the installer) require the revision number. All the outs & sums (that have a git hash inside them) aren't included in the install.

**Comment by sarahcleary** [comment]
Drew 
In addition to name, email - 
We need the following information:
**Licensor name** (as in Company name)  and if an Australian company, ABN or equivalent company registration number 
**Licensor’s Representative**  - ability to insert an additional person to be sent notices/invoices 
**Turnover of Licensor** - this is preferably a tick box – for the turnover of the company/licensor.  (see Pricing structure at bottom of this page - https://www.apsim.info/download-apsim/commercial-use/ )

**Comment by hol430** [comment]
Ok, yeah it sounds like having the release occurring as a separate job after the pull request gets merged will be the way to go. Do we want it to be like ApsimX in that a release is only generated if the pull request fixes an issue (as opposed to working on an issue)?

**Comment by peter-devoil** [comment]
Yes. Hardly anyone shares interim releases.

You could drop the release bits from buildall.xml, and add them to another that just builds the release. Or set the "target" when calling the jobscheduler executable.

**Comment by hol430** [comment]
@peter-devoil @sarahcleary I think the incrementing revision number system is now working. Whenever a pull request is merged, a "ReleaseClassic" job will be triggered on Jenkins iff the pull request fixes an issue. This ReleaseClassic job will figure out the latest revision number, brand and build apsim, and generate an installer. The bob webpage should link to this installer.

There don't appear to be many installers linked from Bob's webpage because it's been a while since someone actually fixed an issue - most of the recent pull requests have all been "working on" issues.

**Comment by peter-devoil** [comment]
Thanks drew.

**Comment by hol430** [comment]
@peter-devoil - do we want to provide Linux sfx/binaries?

**Comment by peter-devoil** [comment]
Not for UQ. @cla473 might use them?

**Comment by sarahcleary** [comment]
@peter-devoil - let me know what needs to go into a news article when you have a chance?  thanks.

**Comment by hol430** [comment]
@sarahcleary @peter-devoil I have published my changes to the registration process:

- Extra options for commercial registrations
- Registering will take you directly to the downloads page (you will still receive an email containing license stuff)

Let me know if you spot anything broken or if you want anything changed

**Comment by peter-devoil** [comment]
Good stuff, drew. My immediate comments:
- we can lose the howXXX (and perfect) stuff from the dropdown menu 
- would be nice if we can embed these pages/iframes in the WP site so users get consistent L&F across the apsim.info domain

**Comment by hol430** [comment]
Ok, I have ditched the howXXX stuff from the dropdown. Regarding integration into the wp site - afaik, the contents of an iframe or embed element aren't affected by the style of the document they're embedded in. Idk much about wordpress though - maybe there is a feature/plugin to work around this.

**Comment by sarahcleary** [comment]
News Article - https://www.apsim.info/no-more-major-apsim-classic-releases/

**Comment by sarahcleary** [comment]
@peter-devoil - can we close this issue?
